src/crimson/osd/shard_services.cc

```diff
@@ -610,15 +610,23 @@ seastar::future<Ref<PG>> ShardServices::make_pg(
 
 seastar::future<Ref<PG>> ShardServices::handle_pg_create_info(
   std::unique_ptr<PGCreateInfo> info) {
-  LOG_PREFIX(OSDSingletonState::trim_maps);
   return seastar::do_with(
     std::move(info),
-    [FNAME, this](auto &info)
+    [this](auto &info)
     -> seastar::future<Ref<PG>> {
       return get_map(info->epoch).then(
-	[&info, FNAME, this](cached_map_t startmap)
+	[&info, this](cached_map_t startmap)
 	-> seastar::future<std::tuple<Ref<PG>, cached_map_t>> {
+	  LOG_PREFIX(ShardServices::handle_pg_create_info);
 	  const spg_t &pgid = info->pgid;
+	  if (!get_map()->is_up_acting_osd_shard(pgid, local_state.whoami)
+	      || !startmap->is_up_acting_osd_shard(pgid, local_state.whoami)) {
```

Comment:
startmap->is_up_acting_osd_shard should always be true.

Reply:
get_map()->is_up_acting_osd_shard(pgid, local_state.whoami) isn't really sufficient -- it will catch a case where there was an interval change and the current map no longer has this OSD in the acting set, but it won't notice if there were two interval changes resulting in this OSD being in the acting set again.

Classic doesn't check this at all.  If we happen to create a PG unnecessarily, it will advance through the remaining maps from its creation epoch, send a message to the new primary, and be removed.  Why is this a particular problem with crimson?

Classic does, however, specifically check the case where the create came from a mon -- see ShardedOpWQ::_process right before the call to handle_pg_create_info -- though it's not clear how that would cause problems either.

Reply:
The issue was that if a pg creation is triggerred by a peering event sent by another osd, all the checks here will be ignore by `by_mon`, and later in this continuation chain, the OSD will get an NULL pool for the pg, which leads to crashes.

Reply:
> The issue was that if a pg creation is triggerred by a peering event sent by another osd and the corresponding pool has been deleted, all the checks here will be ignore by `by_mon`, and later in this continuation chain, the OSD will get an NULL pool for the pg, which leads to crashes.

Reply:
> startmap->is_up_acting_osd_shard should always be true.

The case was like this:

1. `OSD.1` send a peering query for pg `P` to `OSD.2` at osdmap epoch 1000;
2. `OSD.2` respond at osdmap epoch 1500, at which point the pool of  pg `P` has been deleted;
3. `OSD.1` receives `OSD.2`'s response in which the epoch is 1500, and crashes when trying to create pg `P` because it can find the pool for `P` at epoch 1500.

Reply:
> Classic doesn't check this at all. If we happen to create a PG unnecessarily, it will advance through the remaining maps from its creation epoch, send a message to the new primary, and be removed. Why is this a particular problem with crimson?

Classic checks this in `OSD::ShardedOpWQ::_process`: https://github.com/ceph/ceph/blob/5f7baab60b00f662a054f914f27ec19bc2130336/src/osd/OSD.cc#L11085

Reply:
Ah, so it does.